### ISSUE TITLE
Requirements for DOM for use with dynamic control replication

### ISSUE THREAD

**Issue by lightsighter** [StanfordLegion/legion#338]
Here's a list of requirements for changes that need to be made to the existing Soleil DOM code and the Regent compiler before it can be tested with dynamic control replication. Go ahead and check off your boxes when they are done. I'll start testing once all the boxes are checked.

- [x] Regent needs to have the ability to annotate 'replicable' tasks and have that information passed through to the runtime (@elliottslaughter @lightsighter)
- [x] The top-level task in the DOM code needs to be checked that it is 'replicable' and annotated as such (@manopapad @magnatelee)
- [ ] All the major tasks in the iteration loop need to be generating index space task launches with projection region requirements (@manopapad @magnatelee @elliottslaughter)
- [x] The optimization for using the proper subregions for ghost regions needs to incorporated (@sierrakn)
- [x] All partitioning operations need to be done with dependent partitioning operations, e.g. no more colorings (@manopapad @magnatelee)

**Comment by elliottslaughter** [comment]
@lightsighter We already have the first bullet, right?

**Comment by lightsighter** [comment]
@elliottslaughter No, I added support for a static analysis that tries to infer it. This bullet is asking for the case where we explicitly annotate it in the program. This will involve change the Regent parser which I don't know how to do.

**Comment by elliottslaughter** [comment]
@lightsighter You mean `__demand(__replicable)`?

**Comment by lightsighter** [comment]
@elliottslaughter Yes, that's what I mean.

**Comment by elliottslaughter** [comment]
First bullet should be working in 3f7a3ef.

**Comment by lightsighter** [comment]
Yup, first bullet looks good to me.

**Comment by manopapad** [comment]
Bullet 2: For the noobs among us, is there a list of requirements somewhere for what it means to be "replicable"?

Bullet 3: Similarly, can someone explain, for the noobs?

**Comment by elliottslaughter** [comment]
@manopapad For bullet 2 you add `__demand(__replicable)` to your task and see if the compiler yells at you. Note: you will need Legion to be on the `control_replication` branch for this to work.

Similarly for bullet 3 you add `__demand(__parallel)` to any loops where you call tasks, and see if the compiler complains. (Works on any branch.)

**Comment by lightsighter** [comment]
@manopapad I think @elliottslaughter is mostly right about what he told you, but I'm forecasting that the compiler's static analysis is not sufficient yet to be able to safely prove that a task is control replicable so it might give you a false positive error message. In general, in order for a task to be control replicable right now it is not permitted to have any use of a random number generator, it is not allowed to explicitly manipulate phase barriers, and it shouldn't be reading/writing to any files directly. I don't think any of those will be a problem for the DOM top-level task.

**Comment by sierrakn** [comment]
The last two issues are done.

**Comment by lightsighter** [comment]
Thanks @sierrakn! Do you mind pointing me at the code so I can take a look? I want to make sure that the projection functors are going to play nicely with dynamic control replication.

**Comment by manopapad** [comment]
[main DOM code](https://github.com/stanfordhpccenter/soleil-x/blob/master/src/dom.rg)
[standalone DOM harness](https://github.com/stanfordhpccenter/soleil-x/blob/master/src/dom_host.rg)
My understanding is that we don't currently use any non-identity projection functors, and we're not planning to change that, at least not until after the May PSAAP review (Todd's work is not integrated into Legion yet, and even then we will need a compiler extension to use it). I believe the plan is to stick with individual task launches and rely on dynamic tracing to amortize the cost of launching, so bullet 3 is not necessary (@magnatelee may have more to say about this once he gets back).

**Comment by magnatelee** [comment]
@manopapad The code you attached does have non-identity projection functors, not just for sweeps, but also for boundary conditions. Also, if we want to use dynamic control replication for DOM, we need to use index space launches for tasks. Dynamic tracing with individual task launches can be an option when we won't be able to use dynamic control replication for DOM, but it still needs some extension to work with remote mapping. I think you and I will need to find the easiest workaround to make DOM scalable for the review, but the items in this list should still be done to make DOM scalable in the right way.

**Comment by manopapad** [comment]
@elliottslaughter extended the capabilities of Regent's loop optimizer, which allowed us to [launch the sweep tasks in DOM using index-space launches](https://github.com/stanfordhpccenter/soleil-x/blob/wip/dom-index-space-launches/src/dom.rg#L2586-L2596). The [boundary-setting tasks are still launched sequentially](https://github.com/stanfordhpccenter/soleil-x/blob/wip/dom-index-space-launches/src/dom.rg#L2498-L2510), but @magnatelee is fixing this, and also testing the whole thing with DCR.

**Comment by manopapad** [comment]
I just tried to compile Soleil with `__demand(__replicable)` on the work tasks. It's not working because the replicable check is rejecting calls to the following (safe) functions:

* `array`
* helper terra functions in `parallelize_tasks.t` (for Soleil in particular, the replicable check is complaining about the `get_ghost_rect` function on line 173)
* numerical functions in regentlib (`sqrt`, `ceil` etc.)

@elliottslaughter Please add these functions to the whitelist.

**Comment by elliottslaughter** [comment]
Please pull and try again, I think I've got all these....

**Comment by manopapad** [comment]
I am happy to report that the work tasks in Soleil are now accepted by the compiler as replicable. The work tasks in standalone DOM are not (those also use `pow`, which is not accepted by Regent currently -- @elliottslaughter plz fix).

Simply running the resulting code without mapper changes doesn't seem to work (it OOMs on Titan and segfaults on sapling), but that's an issue for another day.

**Comment by lightsighter** [comment]
Only the top-level task should need to be marked as replicable in order for control replication to work.

**Comment by elliottslaughter** [comment]
For @manopapad 's use case, it may plausibly be useful to control replicate multiple levels, since the branching factor at the second level is expected to be quite large. (In fact, if you had to choose one, I think you would choose the second level rather than the first.)

**Comment by lightsighter** [comment]
Ok, that hasn't been tested at all. I would strongly encourage you to make it the top-level task that is control replicated.

**Comment by elliottslaughter** [comment]
Certainly we'll start with that, but it's worth keeping it in mind since it may be performance limiter when we go to scale.

**Comment by manopapad** [comment]
In Soleil the top-level task is just a shim that reads configuration files and launches a number of simulation tasks (usually just one). Each simulation task is assigned a disjoint set of nodes, it creates all regions and partitions them among its nodes, runs all the kernels and outputs results. Each simulation task currently doesn't communicate at all with other simulation tasks. Given this scenario, do you see any issues or inefficiencies with replicating the top-level task?

Regarding the standalone DOM task that I mentioned above, that is a different source file altogether (`src/dom_host.rg` vs `src/soleil.rg`) that only includes the DOM tasks, and was used to debug the DOM solver.

**Comment by elliottslaughter** [comment]
`__demand(__replicable)` is now available in master and `pow` should be accepted. Let me know if there is anything else.

**Comment by manopapad** [comment]
Posting an update on this issue.

- Soleil-X can currently run with and without DCR; at compile time we figure out what your Legion installation supports and compile against that.
- Regent's `__replicable` check succeeds for all work tasks.
- All tasks are launched using index space launches, except DOM sweeps, which are launched individually.
- AFAIK we don't use any non-identity projection functors.

So technically speaking DOM is currently running with DCR, but we would still want to launch all sweep tasks in a single index space launch (with dependencies).

**Comment by elliottslaughter** [comment]
@manopapad Can you give us another update?

**Comment by manopapad** [comment]
I don't have any updates on this. AFAIK the only thing that's missing is the launching of all sweep tasks in a single index-space launch with dependencies, which is not yet supported in the runtime. We currently launch all sweep tasks as sequential tasks. We couls

**Comment by lightsighter** [comment]
I propose we split this out into specific issues that need to be addressed in Legion and Regent and then close this issue. Omnibus issues like this are hard to manage in our new issue tracking framework.

**Comment by elliottslaughter** [comment]
I filed  #715 which I believe is the only outstanding aspect of this issue.
